services/web/server/tests/sandbox/projects_openapi_generator.py

```diff
@@ -0,0 +1,353 @@
+
+    inputs: Dict[InputID, DataSchema]
+    # var inputs?
+    outputs: Dict[OutputID, DataSchema]
+    # var outputs?
+
+
+class Project(BaseModel):
+    id: UUID
+    pipeline: Dict[UUID, Node]
+
+    def update_ids(self, name: str):
+        map_ids: Dict[UUID, UUID] = {}
+        map_ids[self.id] = uuid3(self.id, name)
+        map_ids.update({node_id: uuid3(node_id, name) for node_id in self.pipeline})
+
+        # replace ALL references
+
+
+class Parameter(BaseModel):
+    name: str
+    value: BuiltinTypes
+
+    node_id: UUID
+    output_id: OutputID
+
+
+class Snapshot(BaseModel):
+    id: PositiveInt = Field(..., description="Unique snapshot identifier")
+    label: Optional[str] = Field(None, description="Unique human readable display name")
+    created_at: datetime = Field(
+        default_factory=datetime.utcnow,
+        description="Timestamp of the time snapshot was taken from parent. Notice that parent might change with time",
+    )
+
+    parent_uuid: UUID = Field(..., description="Parent's project uuid")
+    project_uuid: UUID = Field(..., description="Current project's uuid")
+
+
+class ParameterApiModel(Parameter):
+    url: AnyUrl
+    # url_output: AnyUrl
+
+
+class SnapshotApiModel(Snapshot):
+    url: AnyUrl
+    url_parent: AnyUrl
+    url_project: AnyUrl
+    url_parameters: Optional[AnyUrl] = None
+
+    @classmethod
+    def from_snapshot(cls, snapshot: Snapshot, url_for: Callable) -> "SnapshotApiModel":
+        return cls(
+            url=url_for(
+                "get_snapshot",
+                project_id=snapshot.project_id,
+                snapshot_id=snapshot.id,
+            ),
+            url_parent=url_for("get_project", project_id=snapshot.parent_uuid),
+            url_project=url_for("get_project", project_id=snapshot.project_id),
+            url_parameters=url_for(
+                "get_snapshot_parameters",
+                project_id=snapshot.parent_uuid,
+                snapshot_id=snapshot.id,
+            ),
+            **snapshot.dict(),
+        )
+
+
+####################################################################
+
+
+_PROJECTS: Dict[UUID, Project] = {}
+_PROJECT2SNAPSHOT: Dict[UUID, UUID] = {}
+_SNAPSHOTS: Dict[UUID, List[Snapshot]] = defaultdict(list)
+_PARAMETERS: Dict[Tuple[UUID, int], List[Parameter]] = defaultdict(list)
+
+
+####################################################################
+
+
+def get_reverse_url_mapper(request: Request) -> Callable:
+    def reverse_url_mapper(name: str, **path_params: Any) -> str:
+        return request.url_for(name, **path_params)
+
+    return reverse_url_mapper
+
+
+def get_valid_id(project_id: UUID = PathParam(...)) -> UUID:
+    if project_id not in _PROJECTS:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Invalid id")
+    return project_id
+
+
+####################################################################
+
+
+@app.get("/projects/{project_id}", response_model=Project, tags=["project"])
```

Comment:
you might try to explain what you're trying to do here next time... it looks like replacing aiohttp with fastapi?

Reply:
i am using fastapi to prototype with @odeimaiz and even help me producing the OAS document